src/deploy/functions/checkIam.ts

```diff
@@ -104,3 +107,82 @@ export async function checkHttpIam(
   }
   logger.debug("[functions] found setIamPolicy permission, proceeding with deploy");
 }
+
+function reduceEventsToServices(services: Array<Service>, endpoint: backend.Endpoint) {
+  const service = serviceForEndpoint(endpoint);
+  if (!services.find((s) => s.name === service.name)) {
+    services.push(service);
+  }
+  return services;
+}
+
+/**
+ * Checks and sets the roles for specific resource service agents
+ * @param projectId project identifier
+ * @param want backend that we want to deploy
+ * @param have backend that we have currently deployed
+ */
+export async function ensureServiceAgentRoles(
+  projectId: string,
+  want: backend.Backend,
+  have: backend.Backend
+): Promise<void> {
+  // find new services
+  const wantServices = backend.allEndpoints(want).reduce(reduceEventsToServices, []);
+  const haveServices = backend.allEndpoints(have).reduce(reduceEventsToServices, []);
+  const newServices = wantServices.filter(
+    (wantS) => !haveServices.find((haveS) => wantS.name === haveS.name)
+  );
+  if (newServices.length === 0) {
```

Comment:
I know I was in favor of serviceForEndpoint always returning a service (even if noop) but that means that I'll have a getIamPolicy call if I had a pub/sub function and then I add an HTTPS function. I wonder if there's a way to clean this up a bit.

Reply:
I went ahead and added a check in the reduce callback that does not all services that are `noop` or `pubsub` since those don't require us to set bindings

Reply:
Actually having an if check didn't make sense long term, update the type of the `requiredProjectBindings` function to be `undefined` in services and simply check for that in the reduce callback